tests/linearizability/README.md

```diff
@@ -0,0 +1,102 @@
+etcd problem and a bug in testing.
+
+In the end we were inspired by [Jepsen] and implemented a similar set of tests
+that would validate etcd [linearizability]. The framework is able to reproduce
+historical inconsistency issues by inject specific failpoints into etcd server.
+Flakes can be easily distinguished as verification is done on history of
+successful requests and does not depend on failure of any specific request.
+
+[v3.5 data inconsistency postmortem]: https://github.com/etcd-io/etcd/blob/main/Documentation/postmortems/v3.5-data-inconsistency.md#action-items
+[Jepsen]: https://github.com/jepsen-io/jepsen
+[linearizability]: https://cs.brown.edu/~mph/HerlihyW90/p463-herlihy.pdf
+
+
+## How it works
+
+Linearizability tests are not much different from standard e2e tests. They run
+e2e binaries to create a cluster and validate etcd behavior by sending requests
+to API. There are however to major differences. In addition to sending requests
+they also inject different failures into cluster (for example kill etcd process).
+Instead of running any specific test scenario (for example 1. put value 2. get
+value 3. compare if read equals put) they record all the history of requests
+and try to replay them against a simplified model of etcd.
+
+Steps in linearizability tests
+
+1. Setup a etcd cluster
+2. Send traffic to each member and record all requests
+3. Inject failures into the etcd clusters
+4. Validate request history against a etcd model
+
+### Setup a etcd cluster
+
+etcd Linearizability tests use e2e testing framework to setup a local cluster.
+
+### Send traffic to each member and record all requests
+
+What exact requests are sent to etcd is not as important. Traffic can be mostly
+random, however for easier validation it's preferred that:
+* Each write request is unique. Makes tracing requests easier.
+* There are limited number of clients. Cost of validation grows exponentially with number of clients.
+* All clients are running in the same process as requests need to be recorded with accurate time.
+
+### Injecting failures into the etcd cluster
+
+Distributed systems need to be resilient to broad types of failures. To qualify
+etcd we need to explore huge space of possible failure scenarios that can happen.
+Injected failure can be a combinations different failpoints to ensure coverage of
+more complex scenarios.
+
+### Validate request history against a etcd model
+
+Etcd model is a simple struct that implements subset of methods from etcd API used
+to validate correctness of request history. TODO
+
+## Usage
+
+TODO
+
+### Test scenarios
+
+TODO
+
+### How to interpret results from CI
+
+TODO
+
+### Run locally
+
+1. Build etcd with failpoints (can be skipped if you )
+    ```bash
+    make install-gofail
+    FAILPOINTS=true make build
+    ```
+2. Run linearizability tests
+
+    ```bash
+    make test-linearizability
+    ```
+
+### Reading the test results
+
```

Comment:
we should clarify the info that user should submit if they decide to file any issue regarding linearizability test failure that they need help debugging.